openpype/hosts/nuke/api/plugin.py

```diff
@@ -1,27 +1,381 @@
+from collections import OrderedDict, defaultdict
 from abc import abstractmethod
 
-import nuke
+from openpype.client import (
+    get_asset_by_name,
+    get_subsets,
+)
 
 from openpype.settings import get_current_project_settings
+from openpype.lib import (
+    BoolDef,
+    EnumDef
+)
 from openpype.pipeline import (
     LegacyCreator,
     LoaderPlugin,
+    CreatorError,
+    Creator as NewCreator,
+    CreatedInstance,
+    legacy_io
 )
 from .lib import (
+    INSTANCE_DATA_KNOB,
     Knobby,
     check_subsetname_exists,
     maintained_selection,
+    get_avalon_knob_data,
     set_avalon_knob_data,
     add_publish_knob,
     get_nuke_imageio_settings,
     set_node_knobs_from_settings,
+    set_node_data,
+    get_node_data,
     get_view_process_node,
-    get_viewer_config_from_string
+    get_viewer_config_from_string,
+    deprecated
 )
+from .pipeline import (
+    list_instances,
+    remove_instance
+)
+
+
+def _collect_and_cache_nodes(creator):
+    key = "openpype.nuke.nodes"
+    if key not in creator.collection_shared_data:
+        instances_by_identifier = defaultdict(list)
+        for item in list_instances():
+            _, instance_data = item
+            identifier = instance_data["creator_identifier"]
+            instances_by_identifier[identifier].append(item)
+        creator.collection_shared_data[key] = instances_by_identifier
+    return creator.collection_shared_data[key]
+
+
+class NukeCreatorError(CreatorError):
+    pass
+
+
+class NukeCreator(NewCreator):
+    selected_nodes = []
+
+    def pass_pre_attributes_to_instance(
+        self,
+        instance_data,
+        pre_create_data,
+        keys=None
+    ):
+        if not keys:
+            keys = pre_create_data.keys()
+
+        creator_attrs = instance_data["creator_attributes"] = {}
+        for pass_key in keys:
+            creator_attrs[pass_key] = pre_create_data[pass_key]
+
+    def add_info_knob(self, node):
+        if "OP_info" in node.knobs().keys():
+            return
+
+        # add info text
+        info_knob = nuke.Text_Knob("OP_info", "")
+        info_knob.setValue("""
+<span style=\"color:#fc0303\">
+<p>This node is maintained by <b>OpenPype Publisher</b>.</p>
+<p>To remove it use Publisher gui.</p>
+</span>
+        """)
+        node.addKnob(info_knob)
+
+    def check_existing_subset(self, subset_name, instance_data):
+        """Check if existing subset name versions already exists."""
+        # Get all subsets of the current asset
+        project_name = legacy_io.active_project()
+        asset_doc = get_asset_by_name(
+            project_name, instance_data["asset"], fields=["_id"]
+        )
+        subset_docs = get_subsets(
+            project_name, asset_ids=[asset_doc["_id"]], fields=["name"]
+        )
+        existing_subset_names_low = {
+            subset_doc["name"].lower()
+            for subset_doc in subset_docs
+        }
+        return subset_name.lower() in existing_subset_names_low
```

Comment:
This seems like an overly verbose check. If a new nuke script is created after an initial publish and someone intentionally tries to publish into an existing entry they can't create it. Similary if somehow the output node got deleted in their nuke script and they want to recreate it again, they can't. 

Even someone else from another task in the same assets couldn't even try to do so? (or maybe they could if it was prefixed with the task actually; not sure if that's the case).

I feel like a check whether the current Nuke script already has an instance publishing to the same asset + subset name would be a decent validation but the current 'check' seems quite blocking for what can definitely be a production use case.

Reply:
> We should only be checking if a node with particular subset name is not already created in workfile nodes, like you've said.

The next question is - is there ever a case where that might need to be allowed (outside of these creators, outside of the realm of Nuke). If not, then maybe should the publishing system/Creators in general already perform a check like that so that developers don't need to recreate that logic for each individual Creator plugin?

Reply:
I presume this is more question for @iLLiCiTiT

Reply:
To be honest I don't think it's job of creator to check it until it's required for his functionality. This looks like a validation before validators.

Reply:
BTW you can change the context and variant after creation and where the `create` would not be called -> as the validation.